### PR TITLE
[AsyncRL][1/N] Add abort_generation to vllm engine and pause/continue generation to client

### PR DESCRIPTION
**Pull request by CharlieFRuan**
This PR is the first step of supporting in-flight weight updates for Async RL described in literatures like AReal and PipelineRL.

We implement the following:
- `abort_generation` for vLLM engine, which calls `self.engine.abort(...)`, making all in-flight requests return the already completed tokens
- `pause_generation()`, `resume_generation()`, and `_wait_for_generation_to_resume()`, revolving around the control event `generation_paused_event`

Note this does PR does not affect any existing codepath.

We also add unit tests where we test aborting 2 running and 2 waiting requests -- all should return, where latter return zero completion tokens.

Tracked in https://github.com/NovaSky-AI/SkyRL/issues/536

### CODE REVIEW

File: skyrl-train/skyrl_train/inference_engines/inference_engine_client.py [hunk 1]
```diff
@@ -58,6 +62,9 @@ async def _run_on_all_engines(self, method_name: str, *args, **kwargs):
         return await asyncio.gather(*awaitables)
 
     async def generate(self, input_batch: InferenceEngineInput) -> InferenceEngineOutput:
+        assert (
+            not self.generation_paused_event.is_set()
+        ), "pause_generation is unsupported for InferenceEngineClient.generate()."
```
[the remark this answers is not in the document]
Comment (CharlieFRuan) [reply skyrl-train/skyrl_train/inference_engines/inference_engine_client.py]: We will not call `pause_generation()` if users use `generation()` or `completion()` -- it is only supported for `chat_completions()` for now. Does it make sense? @gemini-code-assist

File: skyrl-train/skyrl_train/inference_engines/inference_engine_client.py [hunk 2]
```diff
@@ -149,6 +159,7 @@ async def completion(self, request_payload: Dict[str, Any]) -> Dict[str, Any]:
 
         Regardless, the order will be maintained, i.e. `output["choices"][i]` corresponds to `request["prompt"][i]`.
         """
+        assert not self.generation_paused_event.is_set(), "pause_generation is unsupported for /completions requests."
```
[the remark this answers is not in the document]
Comment (CharlieFRuan) [reply skyrl-train/skyrl_train/inference_engines/inference_engine_client.py]: see above

File: skyrl-train/skyrl_train/inference_engines/inference_engine_client.py [hunk 3]
```diff
@@ -281,6 +292,48 @@ def tp_size(self) -> int:
...
+        2. Then we add a grace period to ensure all in-flight requests have entered the engine's
+           scheduler and hence can be aborted. Otherwise, there can be requests already submitted
+           but not yet entered the scheduler, so we miss to abort them.
+        3. Finally, we call abort requests on all self.engines. This will cause the requests sent from
+           InferenceEngineClient to `InferenceEngineClient.engines` to return the already-generated tokens, but
+           the users' requests to `InferenceEngineClient` will not return but simply blocked.
+        """
+        assert not self.generation_paused_event.is_set(), "Generation is already paused"
+        self.generation_paused_event.set()
+        time.sleep(ABORT_GENERATION_GRACE_PERIOD_SECONDS)
```
[the remark this answers is not in the document]
Comment (CharlieFRuan) [reply skyrl-train/skyrl_train/inference_engines/inference_engine_client.py]: fixed

File: skyrl-train/skyrl_train/inference_engines/inference_engine_client.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (CharlieFRuan) [reply skyrl-train/skyrl_train/inference_engines/inference_engine_client.py]: Fixed, went with suggestion 1

File: skyrl-train/skyrl_train/inference_engines/inference_engine_client.py [hunk 4]
```diff
@@ -281,6 +291,50 @@ def tp_size(self) -> int:
     def dp_size(self) -> int:
         raise NotImplementedError("InferenceEngineClient does not implement dp_size()")
 
+    # ----------------------------
+    # Generation pause and resume
+    # ----------------------------
+    async def _wait_for_generation_to_resume(self) -> None:
+        """Waits for generation to be resumed, intended for in-flight weight updates and partial rollouts."""
+        while self.generation_paused_event.is_set():
+            await asyncio.sleep(0.5)
```
[the remark this answers is not in the document]
Comment (CharlieFRuan) [reply skyrl-train/skyrl_train/inference_engines/inference_engine_client.py:300]: 0.5 is reasonable since it is waiting on weight updates, which takes a long time compared to 500ms

File: skyrl-train/skyrl_train/inference_engines/inference_engine_client.py [hunk 5]
```diff
@@ -281,6 +291,50 @@ def tp_size(self) -> int:
...
+        Currently only supported for `/chat/completions` and not `/completions` or `generate()`.
+
+        Both in-flight and incoming requests will be blocked until `resume_generation` is called.
+        1. We first set the paused event to avoid new requests from being submitted while aborting requests.
+        2. Then we add a grace period to ensure all in-flight requests have entered the engine's
+           scheduler and hence can be aborted. Otherwise, there can be requests already submitted
+           but not yet entered the scheduler, so we miss to abort them.
+        3. Finally, we call abort requests on all self.engines. This will cause the requests sent from
+           InferenceEngineClient to `InferenceEngineClient.engines` to return the already-generated tokens, but
+           the users' requests to `InferenceEngineClient` will not return but simply blocked.
```
[the remark this answers is not in the document]
Comment (CharlieFRuan) [reply skyrl-train/skyrl_train/inference_engines/inference_engine_client.py]: This will be fixed by an incoming PR, as noted in `chat_completion()`'s TODO

File: skyrl-train/skyrl_train/inference_engines/inference_engine_client.py [hunk 6]
```diff
@@ -281,6 +291,52 @@ def tp_size(self) -> int:
...
+        """
+        if self.generation_paused_event.is_set():
+            raise RuntimeError("Generation is already paused, cannot pause again.")
+        self.generation_paused_event.set()
+        await asyncio.sleep(ABORT_GENERATION_GRACE_PERIOD_SECONDS)
```
[the remark this answers is not in the document]
Comment (CharlieFRuan) [reply skyrl-train/skyrl_train/inference_engines/inference_engine_client.py:321]: Good point, we could add a running request counts in `InferenceEngineClient` and make sure we have aborted that many requests. Will leave it as a future TODO. I believe this `sleep()` should account for most scenarios.

File: skyrl-train/skyrl_train/inference_engines/inference_engine_client.py [hunk 7]
```diff
@@ -58,6 +61,8 @@ async def _run_on_all_engines(self, method_name: str, *args, **kwargs):
         return await asyncio.gather(*awaitables)
 
     async def generate(self, input_batch: InferenceEngineInput) -> InferenceEngineOutput:
+        if self.generation_paused_event.is_set():
+            raise RuntimeError("pause_generation is unsupported for InferenceEngineClient.generate().")
```
[the remark this answers is not in the document]
Comment (CharlieFRuan) [reply skyrl-train/skyrl_train/inference_engines/inference_engine_client.py:65]: It is already documented in `pause_generation()`

File: skyrl-train/tests/gpu/gpu_ci/test_pause_and_continue_generation.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+            assert out["choices"][0].get("finish_reason") == "abort"
+
+        # Two requests should have never got to run because we have max_num_seqs=2, and yet they should
+        # be aborted.
+        assert (
+            num_completion_tokens_is_zero == 2
+        ), f"Expected 2 requests with completion_tokens=0, got {num_completion_tokens_is_zero}."
+
+        # Unpause for the next API run
+        client.resume_generation()
```
[the remark this answers is not in the document]
Comment (CharlieFRuan) [reply skyrl-train/tests/gpu/gpu_ci/test_pause_and_continue_generation.py:112]: Will add it in a follow-up PR since we do not support resuming generation yet

**Comment by CharlieFRuan** [comment]
GPU CI running here: https://github.com/NovaSky-AI/SkyRL/actions/runs/18673845719

**Comment by CharlieFRuan** [comment]
ah GPU CI failed due to trying to pickle the thread.Event, ~~fixing~~ fixed, new run here: https://github.com/NovaSky-AI/SkyRL/actions/runs/18701237223

**Comment by CharlieFRuan** [comment]
The new gemini comments are redundant with above rounds

File: skyrl-train/skyrl_train/inference_engines/remote_inference_engine.py [hunk 9]
```diff
@@ -241,6 +241,9 @@ async def _destroy_weights_update_group(self):
             resp = await session.post(f"{self.url}/destroy_weights_update_group")
             return await resp.json()
 
+    async def abort_generation(self) -> None:
+        raise NotImplementedError("Abort generation is not supported for remote inference engines.")
```
Comment (tyler-griggs) [inline skyrl-train/skyrl_train/inference_engines/remote_inference_engine.py:245]: Once you have support on Ray wrapped inference engines, tasks like supporting and testing the remote engine code path are good tasks to create an issue for and hand off to others!

File: skyrl-train/skyrl_train/inference_engines/inference_engine_client.py (hunk 6 above)
Comment (tyler-griggs) [inline skyrl-train/skyrl_train/inference_engines/inference_engine_client.py:321]: I'm pretty suspicious of using a constant here, which can be brittle, but I get why it's used :) A possible alternative is to query vLLM engines until the total number of outstanding requests in the scheduler matches the number of outstanding requests that the inference engine client knows of. But happy to let that come later

File: skyrl-train/tests/gpu/gpu_ci/test_pause_and_continue_generation.py [hunk 10]
```diff
@@ -0,0 +1,112 @@
...
+        model=MODEL,
+        num_inference_engines=cfg.generator.num_inference_engines,
+        sleep_level=1,
+        # We test aborting 2 running requests and 2 waiting requests
+        max_num_seqs=2,
+    )
+
+    tokenizer = AutoTokenizer.from_pretrained(MODEL)
+
+    for api in ["chat_completion", "completion"]:
```
Comment (tyler-griggs) [inline skyrl-train/tests/gpu/gpu_ci/test_pause_and_continue_generation.py:58]: To double check, the reason why you test both `chat_completion` and `completion` here is because we are calling it directly on the engine object (rather than through the inference engine client, which doesn't yet support pause/resume for `completion`), is that right?

File: skyrl-train/tests/gpu/gpu_ci/test_pause_and_continue_generation.py [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+from typing import List
+
+MODEL = "Qwen/Qwen2.5-0.5B-Instruct"
+TP_SIZE = 1
+SERVER_PORT = 8123
+SERVER_HOST = "127.0.0.1"
+
+
+@pytest.mark.vllm
+def test_abort_generation_vllm_engine(ray_init_fixture):
```
Comment (tyler-griggs) [inline skyrl-train/tests/gpu/gpu_ci/test_pause_and_continue_generation.py:22]: Can we add a todo to also test the inference engine client? Namely, this test doesn't cover logic in `chat_completion`, such as calling `_wait_for_generation_to_resume`

File: skyrl-train/skyrl_train/inference_engines/remote_inference_engine.py (hunk 9 above)
Comment (CharlieFRuan) [reply skyrl-train/skyrl_train/inference_engines/remote_inference_engine.py:245]: good point! Added to the tracker

File: skyrl-train/skyrl_train/inference_engines/inference_engine_client.py (hunk 6 above)
Comment (CharlieFRuan) [reply skyrl-train/skyrl_train/inference_engines/inference_engine_client.py:321]: Yep, briefly disccused about it in the gemini comment. Added to the tracker as well

File: skyrl-train/tests/gpu/gpu_ci/test_pause_and_continue_generation.py (hunk 10 above)
Comment (CharlieFRuan) [reply skyrl-train/tests/gpu/gpu_ci/test_pause_and_continue_generation.py:58]: That is correct, I won't support `completion` retry yet. But for pause and continue, we can test both since we directly call `engine`'s function

File: skyrl-train/tests/gpu/gpu_ci/test_pause_and_continue_generation.py (hunk 11 above)
Comment (CharlieFRuan) [reply skyrl-train/tests/gpu/gpu_ci/test_pause_and_continue_generation.py:22]: Assuming the 2/N PR will be merged soon, I will skip the TODO comment here:)
